dial.go

```diff
@@ -308,6 +341,42 @@ func (d FdDialer) Dial(ctx context.Context, opts DialOpts) (Conn, error) {
 	return conn, nil
 }
 
+type AuthDialer struct {
+	// Dialer is a base dialer.
+	Dialer Dialer
+	// Authentication options.
+	Auth     Auth
+	Username string
+	Password string
+}
+
+// Dial makes AuthDialer satisfy the Dialer interface.
+func (d AuthDialer) Dial(ctx context.Context, opts DialOpts) (Conn, error) {
+	conn, err := d.Dialer.Dial(ctx, opts)
+	if err != nil {
+		return conn, err
+	}
+
```

Comment:
Let's add another check. Since with an uninitialized connection, authorization is generally impossible.
```Go
greeting := conn.Greeting()
if greeting == Greeting{} {
     return errors.New("an unexpected connection with initialized greeting")
}
```